frappe/permissions.py

```diff
@@ -361,14 +361,24 @@ def has_user_permission(doc, user=None, debug=False, *, ptype=None):
 		# if allowed_docs is empty it states that there is no applicable permission under the current doctype
 
 		# only check if allowed_docs is not empty
-		if allowed_docs and docname and str(docname) not in allowed_docs:
-			# no user permissions for this doc specified
-			debug and _debug_log(
-				"User doesn't have access to this document because of User Permissions, allowed documents: "
-				+ str(allowed_docs)
-			)
-			push_perm_check_log(_("Not allowed for {0}: {1}").format(_(doctype), docname), debug=debug)
-			return False
+		if allowed_docs:
+			if doc.meta.is_tree and docname is None:
+				from frappe.utils.nestedset import get_ancestors_of
+
+				parent = doc.get(doc.nsm_parent_field)
+				docs_to_check = [parent, *get_ancestors_of(doctype, parent)]
+				condition = any(str(doc) not in allowed_docs for doc in docs_to_check)
```

Comment:
So if I understand this code right, if you have permission on all ancestors, then you can create a new children. This might be excessive requirement? 

Just permission on immediate parent is enough ig. We should allow this this bypass only for `create` and only if the immediate parent has UP that allows descendants? 

@sagarvora can confirm if there are any more loopholes here. :eyes: 


(on side note, this code is getting increasingly complex :skull: )

Reply:
> We should allow this this bypass only for `create` and only if the immediate parent has UP that allows descendants?

Yeah, something like this makes more sense.

Reply:
> Just permission on immediate parent is enough ig

So this won't work for something like A -> B -> C (you have permission on B, and you try to create a doc under C).

Will try to figure out something with ancestors + hide_descendants not enabled